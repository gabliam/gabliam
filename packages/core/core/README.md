[![Build Status][build-image]][build-url]
[![NPM version][npm-image]][npm-url]


# Gabliam core

Gabliam is on top of inversify.
Gabliam help you for register all class in inversify.
Gabilliam is modular with plugins.


# Gabliam phase

## Build phase

1. Load all config file
2. Load all file in scanPath of main
	- All decorators on class add metadata and can register this class
	- if an class has @scanPath, add this folder in queue for load
	- load all plugin
3. Bind classes in DI (inversify)
	- call all plugin.bind
4. load all config classes (order: CoreConfig => PluginConfig => config)
	- call all plugin.config for all instance of config class
5. Build
	- call all plugin.build

## Start phase
/!\ Build phase must be passed

1. call all plugin.start

## Stop phase
/!\ Build and start phases must be passed

1. call all plugin.start

## Detroy phase
/!\ Stop phase must be passed

1. call all plugin.detroy

# Samples
Go to [Samples]



[Samples]: <https://github.com/gabliam/gabliam/tree/master/examples>

# License

  MIT

[build-image]: https://img.shields.io/travis/gabliam/gabliam/master.svg?style=flat-square
[build-url]: https://travis-ci.org/gabliam/gabliam
[npm-image]: https://img.shields.io/npm/v/@gabliam/core.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@gabliam/gabliam/core
