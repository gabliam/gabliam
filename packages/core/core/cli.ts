#!/usr/bin/env node
import 'reflect-metadata';
import yargs from 'yargs';
import { StartCommand } from './src/commands/start-command';

yargs
  .usage('Usage $0 <command> [options]')
  .command(new StartCommand())
  .recommendCommands()
  .demandCommand(1)
  .strict()
  .alias('v', 'version')
  .help('h')
  .alias('h', 'help').argv;

require('yargonaut')
  .style('blue')
  .style('yellow', 'required')
  .helpStyle('green')
  .errorsStyle('red');
