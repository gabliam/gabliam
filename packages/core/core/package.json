{
  "name": "@gabliam/core",
  "version": "0.0.0-PLACEHOLDER",
  "description": "Gabliam is on top of inversify",
  "main": "index.js",
  "typings": "index.d.ts",
  "module": "index.js",
  "scripts": {
    "build": "rimraf lib && tsc -p tsconfig.build.json"
  },
  "files": [
    "lib",
    "src",
    "index.js",
    "index.d.ts"
  ],
  "dependencies": {
    "@gabliam/expression": "0.0.0-PLACEHOLDER",
    "@types/joi": "^14.0.1",
    "caller": "^1.0.1",
    "co": "^4.6.0",
    "debug": "^4.1.1",
    "glob": "^7.1.3",
    "graph-data-structure": "^1.8.0",
    "inversify": "^5.0.1",
    "inversify-logger-middleware": "^3.0.0",
    "joi": "^14.3.1",
    "js-yaml": "^3.12.1",
    "lodash": "^4.17.11",
    "shortstop": "^1.0.3",
    "shortstop-handlers": "^1.0.1",
    "tslib": "^1.9.3"
  },
  "devDependencies": {
    "@types/debug": "^4.1.3",
    "@types/fs-extra": "^5.0.4",
    "@types/glob": "^7.1.1",
    "@types/js-yaml": "^3.11.4",
    "@types/lodash": "^4.14.123",
    "fs-extra": "^7.0.1",
    "properties": "^1.2.1",
    "rxjs": "^6.3.3",
    "toml": "^2.3.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/gabliam/gabliam.git"
  },
  "author": "David Touzet <dtouzet@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/gabliam/gabliam/issues"
  },
  "homepage": "https://github.com/gabliam/gabliam/tree/master/packages/core#readme",
  "engines": {
    "node": ">=8.9.0"
  },
  "gitHead": "ec328fa0b62f474d470f74a2fc9637838f54091e"
}
