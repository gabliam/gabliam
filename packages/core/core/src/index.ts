import 'reflect-metadata';

export {
  decorate,
  inject,
  injectable,
  interfaces as inversifyInterfaces,
  multiInject,
  named,
  optional,
  postConstruct,
  tagged,
  targetName,
  unmanaged,
} from 'inversify';
export * from './common';
export {
  APP_CONFIG,
  CORE_CONFIG,
  INJECT_CONTAINER_KEY,
  VALUE_EXTRACTOR,
} from './constants';
export * from './container';
export * from './decorator';
export * from './errors';
export * from './gabliam';
export * from './gabliam-options';
export * from './gabliam-utils';
export * from './interfaces';
export * from './joi';
export { configResolver, FileLoader, Loader, Resolver } from './loaders';
export * from './metadatas';
export * from './reflection';
export * from './registry';
export * from './value-extractor';
