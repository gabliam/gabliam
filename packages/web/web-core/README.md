[![Build Status][build-image]][build-url]
[![NPM version][npm-image]][npm-url]


# Gabliam web-core

Gabliam plugin for add web-core.
All decorators and utils for web

# Concepts

## ServerStarter

This class is used for start the server. By default, use [HttpServerStarter](./src/server-starter.ts)

## REQUEST_LISTENER_CREATOR

Must return an function that create a [RequestListener](./src/interface.ts#L19)
Use by the serverstarter.


# License

  MIT

[build-image]: https://img.shields.io/travis/gabliam/gabliam/master.svg?style=flat-square
[build-url]: https://travis-ci.org/gabliam/gabliam
[npm-image]: https://img.shields.io/npm/v/@gabliam/web-core.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@gabliam/web-core
