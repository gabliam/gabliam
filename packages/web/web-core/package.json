{
  "name": "@gabliam/web-core",
  "version": "0.0.0-PLACEHOLDER",
  "description": "Gabliam plugin for add web-core",
  "keywords": [
    "gabliam",
    "gabliam-plugin",
    "plugin",
    "web-core",
    "rest"
  ],
  "main": "index.js",
  "typings": "index.d.ts",
  "module": "index.js",
  "scripts": {
    "build": "rimraf lib && tsc -p tsconfig.build.json"
  },
  "files": [
    "lib",
    "src",
    "index.js",
    "index.d.ts"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/gabliam/gabliam.git"
  },
  "author": "David Touzet <dtouzet@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/gabliam/gabliam/issues"
  },
  "homepage": "https://github.com/gabliam/gabliam/tree/master/packages/web/web-core#readme",
  "optionalDependencies": {
    "supertest": "^4.0.2"
  },
  "peerDependencies": {
    "@gabliam/core": "0.0.0-PLACEHOLDER"
  },
  "devDependencies": {
    "@gabliam/core": "0.0.0-PLACEHOLDER",
    "@types/cookies": "^0.7.4",
    "@types/escape-html": "^0.0.20",
    "@types/statuses": "^1.5.0"
  },
  "dependencies": {
    "cookies": "^0.8.0",
    "http-status-codes": "^1.4.0",
    "property-tunnel": "^1.2.2",
    "statuses": "^1.5.0",
    "tslib": "^1.10.0"
  },
  "engines": {
    "node": ">=8.9.0"
  },
  "gitHead": "ec328fa0b62f474d470f74a2fc9637838f54091e"
}
