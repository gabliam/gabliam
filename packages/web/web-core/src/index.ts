export * from './constants';
export * from './cookies';
export * from './errors';
export * from './exceptions';
export * from './execution-context';
export * from './gab-context';
export * from './gab-request';
export * from './gab-response';
export * from './interceptor';
export * from './interface';
export * from './metadatas';
export * from './pipe';
export * from './plugin-config';
export * from './response-entity';
export * from './server-starter';
export * from './utils';
export * from './web-configuration';
export * from './web-plugin-base';
