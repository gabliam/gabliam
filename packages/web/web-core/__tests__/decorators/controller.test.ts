import { reflection } from '@gabliam/core';
import { Controller, RestController } from '@gabliam/web-core';

describe('Controller decorator', () => {
  test('should add Controller metadata to a class when decorated with @Controller(string)', async () => {
    @Controller('/test')
    class TestController {}

    expect(reflection.annotations(TestController)).toMatchSnapshot();
  });

  test('should add Controller metadata to a class when decorated with @Controller(ControllerOptions)', async () => {
    @Controller({
      name: 'TestControllerName',
      path: '/test',
    })
    class TestController {}

    expect(reflection.annotations(TestController)).toMatchSnapshot();
  });

  test('should fail when decorated multiple times the same class with @Controller', () => {
    expect(function() {
      @Controller({
        name: 'TestControllerName',
        path: '/test2',
      })
      @Controller('/test')
      class TestConfig {}
      new TestConfig();
    }).toThrowError();
  });
});

describe('RestController decorator', () => {
  test('should add Controller metadata to a class when decorated with @RestController(string)', async () => {
    @RestController('/test')
    class TestController {}

    expect(reflection.annotations(TestController)).toMatchSnapshot();
  });

  test('should add Controller metadata to a class when decorated with @RestController(ControllerOptions)', async () => {
    @RestController({
      name: 'TestControllerName',
      path: '/test',
    })
    class TestController {}

    expect(reflection.annotations(TestController)).toMatchSnapshot();
  });

  test('should fail when decorated multiple times the same class with @RestController', () => {
    expect(function() {
      @RestController({
        name: 'TestControllerName',
        path: '/test2',
      })
      @RestController('/test')
      class TestConfig {}
      new TestConfig();
    }).toThrowError();
  });
});
