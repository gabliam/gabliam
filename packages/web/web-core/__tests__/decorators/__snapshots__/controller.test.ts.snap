// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controller decorator should add Controller metadata to a class when decorated with @Controller(ControllerOptions) 1`] = `
[
  DecoratorFactory {
    "json": false,
    "name": "TestControllerName",
    "path": "/test",
  },
  DecoratorFactory {
    "autobind": true,
    "id": "TestControllerName",
    "type": "ControllerType",
  },
]
`;

exports[`Controller decorator should add Controller metadata to a class when decorated with @Controller(string) 1`] = `
[
  DecoratorFactory {
    "json": false,
    "name": undefined,
    "path": "/test",
  },
  DecoratorFactory {
    "autobind": true,
    "id": [Function],
    "type": "ControllerType",
  },
]
`;

exports[`RestController decorator should add Controller metadata to a class when decorated with @RestController(ControllerOptions) 1`] = `
[
  DecoratorFactory {
    "json": true,
    "name": "TestControllerName",
    "path": "/test",
  },
  DecoratorFactory {
    "autobind": true,
    "id": "TestControllerName",
    "type": "ControllerType",
  },
]
`;

exports[`RestController decorator should add Controller metadata to a class when decorated with @RestController(string) 1`] = `
[
  DecoratorFactory {
    "json": true,
    "name": undefined,
    "path": "/test",
  },
  DecoratorFactory {
    "autobind": true,
    "id": [Function],
    "type": "ControllerType",
  },
]
`;
